DatadogRUM/Sources/Instrumentation/AppHangs/AppHangsWatchdogThread.swift

```diff
@@ -0,0 +1,131 @@
+    /// Closure to be notified when App Hang ends. It will be executed on the watchdog thread.
+    @ReadWriteLock
+    internal var onHangEnded: ((AppHang) -> Void)?
+
+    /// Creates an instance of an App Hang watchdog thread.
+    ///
+    /// This thread is not started by default and requires manual invocation of `.start()`.
+    ///
+    /// - Parameters:
+    ///   - appHangThreshold: Minimum duration of the `queue` hang to consider it an App Hang.
+    ///   - queue: The queue to observe for hangs. (main queue)
+    ///   - dateProvider: Date provider.
+    ///   - telemetry: The handler to report issues through RUM Telemetry.
+    init(
+        appHangThreshold: TimeInterval,
+        queue: DispatchQueue,
+        dateProvider: DateProvider,
+        telemetry: Telemetry
+    ) {
+        self.appHangThreshold = appHangThreshold
+        self.idleInterval = appHangThreshold * Constants.tolerance
+        self.mainQueue = queue
+        self.dateProvider = dateProvider
+        self.telemetry = telemetry
+        super.init()
+        self.name = "com.datadoghq.app-hang-watchdog"
+    }
+
+    override func main() {
+        let mainThreadTask = self.mainThreadTask
+
+        while !isCancelled {
+            defer {
+                // Before continuing, sleep for a while, to reduce the CPU consumption by watchdog thread.
+                Thread.sleep(forTimeInterval: idleInterval)
+            }
+
+            let waitStart: DispatchTime = .now()
+
+            // Schedule task on the main thread to measure how fast it responds
+            mainQueue.async {
+                mainThreadTask.signal()
+            }
+
+            // Await task completion
+            let result = mainThreadTask.wait(timeout: waitStart + appHangThreshold)
+
+            if result == .success {
+                // This is predominant case of "no hang" situation. The main thread executed the task way below hang threshold.
+                continue
+            } // Otherwise, `result == .timeOut`, meaning that the main thread task is delayed to more than hang threshold.
+
+            // We expect to be here roughtly `~appHangThreshold` after `wait()`. If this code is executed well after (50% margin),
+            // assume that the thread was suspended and app just woke up. In such case, ignore the hang as likely false-positive.
+            if interval(from: waitStart, to: .now()) > appHangThreshold * 1.5 {
+                DD.logger.debug("Ignoring likely false-positive App Hang")
+                continue // ignore likely false-positive
+            }
+
+            // TODO: RUM-2925 Capture stack trace of the pending App Hang
+
+            // Previous wait timed out, so wait again for the task completion, this time infinitely until the hang ends.
+            mainThreadTask.wait()
+
+            // The hang has finished.
+            let hangDuration = interval(from: waitStart, to: .now())
+
+            if hangDuration > 30 { // sanity-check
```

Comment:
Because it's "roughly" 10s we may cause false-negatives and actually lost some real hangs. I really want to see some false-positive examples in telemetry, so we can adjust it better, based on real data.